packages/react-server/core/renderMiddleware.js

```diff
@@ -105,16 +105,25 @@ module.exports = function(server, routes) {
 			navigateDfd.resolve();
 
 			if (err) {
-				logger.log("onNavigate received a non-2xx HTTP code", err);
-				handleResponseComplete(req, res, context, start, page);
-				if (err.status && err.status === 404) {
-					next();
-				} else if (err.status === 301 || err.status === 302 || err.status === 307) {
+				// The page can elect to proceed to render
+				// even with a non-2xx response.  If it
+				// _doesn't_ do so then we're done.
+				var done = !(page && page.getHaveDocument());
+
+				if (err.status === 301 || err.status === 302 || err.status === 307) {
 					res.redirect(err.status, err.redirectUrl);
-				} else {
-					next(err);
+				} else if (done) {
```

Comment:
It looks like the previous behavior was to issue `res.redirect`, and then `return;` since the return statement moved inside `if (done)` down below, it's not obvious that that's the case anymore. (In any event, I was under the impression 301, 302 aren't supposed to have additional content?)